### PR TITLE
Remove custom Hermes config for Android 

### PR DESCRIPTION
**Pull request by Ashoat**
## Summary

Right now, `react-native` on Android passes in an explicit Hermes config when initializes Hermes. The upcoming version of Hermes shipping with React Native 0.66 will handle this default config itself, so there's no need to override it from Android anymore.

## Test Plan

I compiled and ran a React Native app using the Android build, making sure to build `ReactAndroid` from source. I confirmed that the config was actually being applied by testing how much memory an application could eat before being killed.

### CODE REVIEW

**Comment by Huxpro** [comment]
Cite my comments at https://github.com/facebook/hermes/issues/511#issuecomment-887929120

> I'm wondering would it make more sense to just increase the default number within Hermes and remove the `React-Android` specific `GCConfig` in React Native instead? As more platforms (Windows, macOS) adopting Hermes, it seems suboptimal to glue this default GCConfig again and again.

**Comment by Ashoat** [comment]
Okay, with that approach in mind I think we can break down the work into three parts:

1. Right now `react-native` always passes in a config when it calls `hermes::makeHermesRuntime`. We should update it so it doesn't pass in a config unless there is a custom one being specified. That way, Hermes will be able to determine a default config on its own.
2. While iOS does not bother specifying a custom config currently, Android does. If we want both platforms to defer to Hermes for the default config, we should remove the custom Android config.
3. Finally, Hermes need to be updated to specify the default config that is currently being used for Android. (This change should be synced with change 2.)

This PR has been updated to handle 1. I can submit a follow-up PR to handle 2, but I'll need some help with 3 since I'm not familiar with how to compile and test a custom Hermes build on iOS and Android. If anybody has any pointers that would be great. Alternately, would be happy if somebody else offers to handle 3 :)

**Comment by Ashoat** [comment]
Rebased and tests are all looking clear! I think this is ready for review/merge. Adding some folks:

- @Huxpro on the Hermes side
- @alloy who wrote `HermesExecutorFactory.cpp`, and @TheSavior who reviewed it
- @grabbou also offered to review on Discord :)

Regarding the changelog – I omitted one because this is because a refactor, and should be a no-op on its own. Can add one if people think it's necessary.

**Review by Huxpro — CHANGES_REQUESTED** [review]
> 1. Right now react-native always passes in a config when it calls hermes::makeHermesRuntime. We should update it so it doesn't pass in a config unless there is a custom one being specified. That way, Hermes will be able to determine a default config on its own.

I see where you were coming from, but I don't see this to be particularly useful because:
1. the `runtimeConfig` is already an optional parameter so `HermesExecutorFactory` is effectively overloaded as what you changed to.
2. the `runtimeConfig` is already default constructed for the case of 2-params overloads so it's effectively the same as deferring it to the `hermes.h`.
3. the use of `folly::Optional` doesn't make anything safer.

> 2. While iOS does not bother specifying a custom config currently, Android does. If we want both platforms to defer to Hermes for the default config, we should remove the custom Android config.

How do you feel about we repurpose this PR to just remove the custom Android config? The Hermes default config change should be committed on Monday.

**Comment by Ashoat** [comment]
Ah, I see what you mean... `makesHermesRuntime`'s default constructor [gets a default-initialized](https://github.com/facebook/hermes/blob/main/API/hermes/hermes.h#L239-L241) `RuntimeConfig` anyways. This PR (as it stands) was built on the premise that these are different:

```js
hermes::makeHermesRuntime(::hermes::vm::RuntimeConfig());
hermes::makeHermesRuntime();
```

But I guess that's not the case. If there's no plan to differentiate these, then I can do as you suggest and make this PR just about reverting the custom Android config.

**Comment by Ashoat** [comment]
Sorry this has been taking so long! My life has been a bit crazy recently and I haven't been able to find the time.

Looking at this now :)

**Comment by Ashoat** [comment]
Okay, the PR has been updated!

**Comment by yungsters** [comment]
Heads up, @Huxpro flagged concerns with this change to me. I just submitted a revert, and he's planning to follow up with more details and proposal for next steps.

File: ReactAndroid/src/main/java/com/facebook/hermes/reactexecutor/OnLoad.cpp [hunk 1]
```diff
@@ -30,26 +28,6 @@ static void hermesFatalHandler(const std::string &reason) {
...
-static ::hermes::vm::RuntimeConfig makeRuntimeConfig(jlong heapSizeMB) {
-  namespace vm = ::hermes::vm;
-  auto gcConfigBuilder =
-      vm::GCConfig::Builder()
-          .withName("RN")
-          // For the next two arguments: avoid GC before TTI by initializing the
-          // runtime to allocate directly in the old generation, but revert to
-          // normal operation when we reach the (first) TTI point.
-          .withAllocInYoung(false)
-          .withRevertToYGAtTTI(true);
```
Comment (Huxpro) [inline ReactAndroid/src/main/java/com/facebook/hermes/reactexecutor/OnLoad.cpp:42]: Note that this is a flip of [the default GCConfig](https://github.com/facebook/hermes/blob/5f2b47d0be6281fd2605d24efc0b43af42b4033d/public/hermes/Public/GCConfig.h#L185-L188) that we customized to optimize TTI for all React Native apps but not going to work on any other environments (hence not suitable to be the Hermes default).

**Review by Huxpro — COMMENTED** [review]
Thanks @yungsters and @Ashoat !

Regarding the decision of a revert...In a nutshell, I felt like you went too far on removing stuffs that we actually would like to keep. 

You removed some critical GCConfig customization (as I highlighted below) which will regress TTI significantly. I'm sorry if my previous wording of ["remove the React-Android specific GCConfig in React Native instead"](https://github.com/facebook/hermes/issues/511#issuecomment-887929120) was a little misleading. What I really meant is reverting just the https://github.com/facebook/react-native/commit/63d20d3b1ef35cb4398d63d62f631f7f5d2935c7 as you pointed out as ["On Android it appears to be overriden to 1024 MiB"](https://github.com/facebook/hermes/issues/511#issuecomment-883185729).

In addition, even if we want to remove all the customized GCConfig, there are [way more RuntimeConfig items other than GCConfig](https://github.com/facebook/hermes/blob/5f2b47d0be6281fd2605d24efc0b43af42b4033d/public/hermes/Public/RuntimeConfig.h) so we probably don't want to remove all the RuntimeConfig threading and JNI scaffolding. I understand where you are coming from because GC heapSize happened to be the only POJO field here and rendered them like dead code. However: 
1. It's a moving part that we'll constantly modify whenever we felt the need to expose some Hermes config (e.g. We was using it for Proxy. Who knows what's next?).
2. A unknown fun fact is that we actually have an internal mirror of this file at FB that exposed way more Hermes config to Java and we'd love to keep the OSS and internal interface consistent.

Due to the [tight timeline we have for RN 0.66 (which is already cut and is waiting Hermes for RC2 release)](https://github.com/react-native-community/releases/issues/246), I have to revert it myself at https://github.com/facebook/react-native/commit/a8ac5ea32676dedb17b63069a0c83ddf71746888 to make sure things can be cherry-picked on time. 

Big shout to all the investigation and effort you've made and helped us resolving https://github.com/facebook/hermes/issues/511 and similar issue once and for all!

**Comment by Ashoat** [comment]
No worries – I'm glad to have helped!

**Pull request closed without merging**